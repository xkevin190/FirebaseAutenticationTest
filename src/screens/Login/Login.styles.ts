import { StyleSheet, ViewStyle, ImageStyle, TextStyle } from 'react-native';

import { COLORS } from '../../constants/styles';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    flexDirection: 'column',
    backgroundColor: COLORS.GRAY_LIGHT,
  } as ViewStyle,
  opyaLogo: {
    width: '50%',
    height: '20%',
  } as ImageStyle,
  inputsContainer: {
    width: '100%',
    paddingHorizontal: 24,
    marginBottom: 8,
  } as ViewStyle,
  subtitleContainer: {
    marginBottom: 40,
  } as TextStyle,
  signUpButton: {
    marginTop: 32,
  } as ViewStyle,
  forgotPasswordContainer: {
    alignSelf: 'flex-start',
    paddingHorizontal: 40,
    marginBottom: 30,
  } as ViewStyle,
});

export default styles;
