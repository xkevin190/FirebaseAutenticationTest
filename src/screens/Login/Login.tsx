import React, {useState, useEffect} from 'react';
import {
  View,
  KeyboardAvoidingView,
  TouchableWithoutFeedback,
  Keyboard,
} from 'react-native';
import {useNavigation, useFocusEffect} from '@react-navigation/native';
import {useTranslation} from 'react-i18next';
import {SIGN_UP, FORGOT_PASSWORD} from '../../navigation/routes';
import {Button, CustomText, CustomInput, HyperText} from '../../shared';

import {useNavigationRootStack} from '../../types/navigation';
import styles from './Login.styles';
import {COLORS} from '../../constants/styles';
import Controller from '../../shared/Controller';
import {useLoginForm} from './hooks/useLoginForm';
import {SignInRequest} from '../../types/FirebaseService';
import {authActions, signInThunk} from '../../store/Auth/slice';
import useAppDispatch from '../../hooks/useAppDispatch';
import {authLoading, errorLogin} from '../../store/Auth/selectors';
import useAppSelector from '../../hooks/useAppSelector';

const ControlledInputs = Controller(CustomInput);

const Login: React.FC = () => {
  const [isPasswordHidden] = useState(true);
  const navigation = useNavigation<useNavigationRootStack>();
  const dispatch = useAppDispatch();
  const error = useAppSelector(errorLogin);

  const loading = useAppSelector(authLoading);
  const {t} = useTranslation();

  const {control, watch, handleSubmit, setError, clearErrors} = useLoginForm({
    email: '',
    password: '',
  });

  const signIn = (data: SignInRequest) => {
    dispatch(authActions.resetError());
    dispatch(signInThunk(data));
  };

  useFocusEffect(
    React.useCallback(() => {
      clearErrors();
      return () => {
        dispatch(authActions.resetError());
      };
    }, []),
  );

  useEffect(() => {
    if (error) {
      setError('email', {
        type: 'manual',
        message: '',
      });

      setError('password', {
        type: '',
        message: t(`signUp:translation.signUpError.${error}`),
      });
    }
  }, [error]);

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
      <View style={styles.container}>
        <CustomText
          bodyType="large-paragraph"
          center
          color={COLORS.OPYA_BLUE}
          text={t('login:translation.title')}
          customStyle={styles.subtitleContainer}
        />
        <KeyboardAvoidingView style={styles.inputsContainer} behavior="padding">
          <ControlledInputs
            name="email"
            control={control}
            value={watch('email')}
            title={t('login:translation.emailInput.label')}
            placeholder={t('login:translation.emailInput.placeholder')}
          />
          <ControlledInputs
            name="password"
            control={control}
            value={watch('password')}
            title={t('login:translation.passwordInput.label')}
            placeholder={t('login:translation.passwordInput.placeholder')}
            password={isPasswordHidden}
          />
        </KeyboardAvoidingView>
        <HyperText
          onPress={() => navigation.navigate(FORGOT_PASSWORD)}
          customStyles={styles.forgotPasswordContainer}
          text={t('login:translation.forgotPasswordButton')}
          testID={''}
        />
        <Button
          buttonText={t(
            `login:translation.loginButton.${loading ? 'loading' : 'label'}`,
          )}
          textColor={COLORS.WHITE}
          rounded
          disabled={loading}
          onPress={handleSubmit(signIn)}
          testID={'loginButton'}
        />
        <HyperText
          onPress={() => navigation.navigate(SIGN_UP)}
          customStyles={styles.signUpButton}
          prefix={t('login:translation.signUp.prefix')}
          text={t('login:translation.signUp.label')}
          testID={''}
        />
      </View>
    </TouchableWithoutFeedback>
  );
};

export default Login;
