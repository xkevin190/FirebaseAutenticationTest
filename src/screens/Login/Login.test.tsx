import React from 'react';
import {
  render,
  fireEvent,
  RenderResult,
  waitFor,
} from '@testing-library/react-native';
import {NavigationContext} from '@react-navigation/native';
import Login from './Login';

let mockDispatch = jest.fn();
jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');

jest.mock('../../hooks/useAppDispatch',  () => {
    return ()=> mockDispatch;
});

jest.mock('../../hooks/useAppSelector', () => jest.fn());
jest.mock('../../store/Auth/selectors', () => jest.fn());

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));

jest.mock('../../store/Auth/slice', () => {
  return {
    createAccountThunk: jest.fn(),
    authActions: {
      resetError: jest.fn(),
      resetSucess: jest.fn(),
      clearErrors: jest.fn(),
    },
  };
});

const navContext = {
    isFocused: () => true,
  
    addListener: jest.fn(() => jest.fn()),
  };

let Component: RenderResult;

beforeEach(() => {
  Component = render(
    <NavigationContext.Provider value={navContext}>
      <Login />
    </NavigationContext.Provider>,
  );
});


// describe('SignUp component',  () => {
//   test('renders all input fields',  () => {
//     const {getByPlaceholderText, toJSON} = Component;
//     const firstNameInput = getByPlaceholderText(
//       'signUp:translation.nameInput.placeholder',
//     );
//     const lastNameInput = getByPlaceholderText(
//       'signUp:translation.lastNameInput.placeholder',
//     );
//     const emailInput = getByPlaceholderText(
//       'signUp:translation.emailInput.placeholder',
//     );
//     const passwordInput = getByPlaceholderText(
//       'signUp:translation.passwordInput.placeholder',
//     );

//     expect(toJSON()).toMatchSnapshot();
//     expect(firstNameInput).toBeTruthy();
//     expect(lastNameInput).toBeTruthy();
//     expect(emailInput).toBeTruthy();
//     expect(passwordInput).toBeTruthy();
//   });

//   test('should writte in all input fields', async () => {
//     const {getByPlaceholderText, getByTestId} = Component;
//     const firstNameInput = getByPlaceholderText(
//       'signUp:translation.nameInput.placeholder',
//     );
//     const lastNameInput = getByPlaceholderText(
//       'signUp:translation.lastNameInput.placeholder',
//     );
//     const emailInput = getByPlaceholderText(
//       'signUp:translation.emailInput.placeholder',
//     );
//     const passwordInput = getByPlaceholderText(
//       'signUp:translation.passwordInput.placeholder',
//     );
//     const submitButton = getByTestId('createAccountButton');


//     fireEvent.changeText(firstNameInput, 'John');
//     fireEvent.changeText(lastNameInput, 'Doe');
//     fireEvent.changeText(emailInput, 'test@gmail.com');
//     fireEvent.changeText(passwordInput, 'K123456v?');


//     expect(firstNameInput.props.value).toBe('John');
//     expect(lastNameInput.props.value).toBe('Doe');
//     expect(emailInput.props.value).toBe('test@gmail.com');
//     expect(passwordInput.props.value).toBe('K123456v?');


//     await waitFor(() => {
//         fireEvent.press(submitButton) 
//     });
//     expect(mockDispatch).toHaveBeenCalled();
    
//   });

//   test('should show error in all inputs', async () => {
//     const {getByPlaceholderText, debug, getByTestId, queryAllByTestId, update} =
//       Component;
//     const firstNameInput = getByPlaceholderText(
//       'signUp:translation.nameInput.placeholder',
//     );
//     const lastNameInput = getByPlaceholderText(
//       'signUp:translation.lastNameInput.placeholder',
//     );
//     const emailInput = getByPlaceholderText(
//       'signUp:translation.emailInput.placeholder',
//     );
//     const passwordInput = getByPlaceholderText(
//       'signUp:translation.passwordInput.placeholder',
//     );
//     const submitButton = getByTestId('createAccountButton');

//     fireEvent.changeText(firstNameInput, '');
//     fireEvent.changeText(lastNameInput, '');
//     fireEvent.changeText(emailInput, '');
//     fireEvent.changeText(passwordInput, '');
//     fireEvent.press(submitButton);

//     await waitFor(() => {
//       expect(queryAllByTestId('helperText')).toHaveLength(4);
//     });
    
//   });
// });
