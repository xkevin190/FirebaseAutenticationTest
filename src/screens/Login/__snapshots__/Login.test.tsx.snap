// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders all input fields 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F6F8FF",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      {
        "color": "#2162E3",
        "fontFamily": "Lato-Regular",
        "fontSize": 20,
        "marginBottom": 40,
        "textAlign": "center",
      }
    }
    testID=""
  >
    login:translation.title
  </Text>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "marginBottom": 8,
          "paddingHorizontal": 24,
          "width": "100%",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#9DA5B9",
          "fontFamily": "Lato-Bold",
          "fontSize": 12,
        }
      }
      testID=""
    >
      LOGIN:TRANSLATION.EMAILINPUT.LABEL
    </Text>
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#9DA5B9",
          "borderRadius": 16,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginBottom": 18,
          "marginTop": 6,
          "width": "100%",
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        control={
          {
            "_defaultValues": {
              "email": "",
              "password": "",
            },
            "_disableForm": [Function],
            "_executeSchema": [Function],
            "_fields": {
              "email": {
                "_f": {
                  "mount": true,
                  "name": "email",
                  "ref": {
                    "name": "email",
                  },
                  "value": "",
                },
              },
              "password": {
                "_f": {
                  "mount": true,
                  "name": "password",
                  "ref": {
                    "name": "password",
                  },
                  "value": "",
                },
              },
            },
            "_formState": {
              "dirtyFields": {},
              "disabled": false,
              "errors": {},
              "isDirty": false,
              "isLoading": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "submitCount": 0,
              "touchedFields": {},
              "validatingFields": {},
            },
            "_formValues": {
              "email": "",
              "password": "",
            },
            "_getDirty": [Function],
            "_getFieldArray": [Function],
            "_getWatch": [Function],
            "_names": {
              "array": Set {},
              "mount": Set {
                "email",
                "password",
              },
              "unMount": Set {},
              "watch": Set {
                "email",
                "password",
              },
            },
            "_options": {
              "defaultValues": {
                "email": "",
                "password": "",
              },
              "delayError": 100,
              "mode": "onBlur",
              "reValidateMode": "onChange",
              "resolver": [Function],
              "shouldFocusError": true,
            },
            "_proxyFormState": {
              "dirtyFields": false,
              "disabled": true,
              "errors": true,
              "isDirty": false,
              "isValid": false,
              "isValidating": false,
              "touchedFields": false,
              "validatingFields": false,
            },
            "_removeUnmounted": [Function],
            "_reset": [Function],
            "_resetDefaultValues": [Function],
            "_setErrors": [Function],
            "_state": {
              "action": false,
              "mount": true,
              "watch": false,
            },
            "_subjects": {
              "array": {
                "next": [Function],
                "observers": [],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
              "state": {
                "next": [Function],
                "observers": [
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                ],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
              "values": {
                "next": [Function],
                "observers": [
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                ],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
            },
            "_updateDisabledField": [Function],
            "_updateFieldArray": [Function],
            "_updateFormState": [Function],
            "_updateValid": [Function],
            "getFieldState": [Function],
            "handleSubmit": [Function],
            "register": [Function],
            "setError": [Function],
            "unregister": [Function],
          }
        }
        multiline={false}
        name="email"
        numberOfLines={1}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="login:translation.emailInput.placeholder"
        placeholderTextColor="#9DA5B9"
        secureTextEntry={false}
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontFamily": "Lato-Regular",
            "fontSize": 14,
            "paddingHorizontal": 16,
            "paddingVertical": 10,
          }
        }
        textAlignVertical="auto"
        value=""
      />
    </View>
    <Text
      style={
        {
          "color": "#9DA5B9",
          "fontFamily": "Lato-Bold",
          "fontSize": 12,
        }
      }
      testID=""
    >
      LOGIN:TRANSLATION.PASSWORDINPUT.LABEL
    </Text>
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#9DA5B9",
          "borderRadius": 16,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginBottom": 18,
          "marginTop": 6,
          "width": "100%",
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        control={
          {
            "_defaultValues": {
              "email": "",
              "password": "",
            },
            "_disableForm": [Function],
            "_executeSchema": [Function],
            "_fields": {
              "email": {
                "_f": {
                  "mount": true,
                  "name": "email",
                  "ref": {
                    "name": "email",
                  },
                  "value": "",
                },
              },
              "password": {
                "_f": {
                  "mount": true,
                  "name": "password",
                  "ref": {
                    "name": "password",
                  },
                  "value": "",
                },
              },
            },
            "_formState": {
              "dirtyFields": {},
              "disabled": false,
              "errors": {},
              "isDirty": false,
              "isLoading": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": false,
              "isValidating": false,
              "submitCount": 0,
              "touchedFields": {},
              "validatingFields": {},
            },
            "_formValues": {
              "email": "",
              "password": "",
            },
            "_getDirty": [Function],
            "_getFieldArray": [Function],
            "_getWatch": [Function],
            "_names": {
              "array": Set {},
              "mount": Set {
                "email",
                "password",
              },
              "unMount": Set {},
              "watch": Set {
                "email",
                "password",
              },
            },
            "_options": {
              "defaultValues": {
                "email": "",
                "password": "",
              },
              "delayError": 100,
              "mode": "onBlur",
              "reValidateMode": "onChange",
              "resolver": [Function],
              "shouldFocusError": true,
            },
            "_proxyFormState": {
              "dirtyFields": false,
              "disabled": true,
              "errors": true,
              "isDirty": false,
              "isValid": false,
              "isValidating": false,
              "touchedFields": false,
              "validatingFields": false,
            },
            "_removeUnmounted": [Function],
            "_reset": [Function],
            "_resetDefaultValues": [Function],
            "_setErrors": [Function],
            "_state": {
              "action": false,
              "mount": true,
              "watch": false,
            },
            "_subjects": {
              "array": {
                "next": [Function],
                "observers": [],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
              "state": {
                "next": [Function],
                "observers": [
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                ],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
              "values": {
                "next": [Function],
                "observers": [
                  {
                    "next": [Function],
                  },
                  {
                    "next": [Function],
                  },
                ],
                "subscribe": [Function],
                "unsubscribe": [Function],
              },
            },
            "_updateDisabledField": [Function],
            "_updateFieldArray": [Function],
            "_updateFormState": [Function],
            "_updateValid": [Function],
            "getFieldState": [Function],
            "handleSubmit": [Function],
            "register": [Function],
            "setError": [Function],
            "unregister": [Function],
          }
        }
        multiline={false}
        name="password"
        numberOfLines={1}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="login:translation.passwordInput.placeholder"
        placeholderTextColor="#9DA5B9"
        secureTextEntry={true}
        style={
          {
            "color": "#000000",
            "flex": 1,
            "fontFamily": "Lato-Regular",
            "fontSize": 14,
            "paddingHorizontal": 16,
            "paddingVertical": 10,
          }
        }
        textAlignVertical="auto"
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignSelf": "flex-start",
        "marginBottom": 30,
        "opacity": 1,
        "paddingHorizontal": 40,
      }
    }
    testID=""
  >
    <Text>
      <Text
        style={
          {
            "color": "#2162E3",
            "fontFamily": "Lato-Bold",
            "fontSize": 14,
          }
        }
        testID=""
      >
        login:translation.forgotPasswordButton
      </Text>
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2162E3",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "width": "45%",
      }
    }
    testID="loginButton"
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "Lato-Regular",
          "fontSize": 14,
        }
      }
      testID=""
    >
      login:translation.loginButton.label
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "marginTop": 32,
        "opacity": 1,
      }
    }
    testID=""
  >
    <Text>
      <Text
        style={
          {
            "color": "#9DA5B9",
            "fontFamily": "Lato-Regular",
            "fontSize": 14,
          }
        }
        testID=""
      >
        login:translation.signUp.prefix 
      </Text>
      <Text
        style={
          {
            "color": "#2162E3",
            "fontFamily": "Lato-Bold",
            "fontSize": 14,
          }
        }
        testID=""
      >
        login:translation.signUp.label
      </Text>
    </Text>
  </View>
</View>
`;
